### ISSUE TITLE
Add a Dynamic Sitemap example that works with the new Dynamic Routes

### ISSUE THREAD

**Issue by creativiii** [vercel/next.js#9051]
# Example request

Up until now I've been using a custom Express server with next-routes to handle all my routing. I'm very excited to upgrade and finally move to the Dynamic Routing that was introduced with NextJS 9.

That said, a key part of my site is the Sitemap. At the moment it's very easy for me to have one thanks to my custom `server.js` and the [sitemap package](https://www.npmjs.com/package/sitemap#example-of-using-sitemapjs-with-express) however I am completely lost on how to achieve this with dynamic routing.

I'm not sure if this would be considered out of scope for NextJS, but Sitemaps are essential for websites and its unfortunate that deciding to use a great feature like Dynamic Routing means that I'd have no obvious way of creating one.

An example showing how to use the existing Sitemap NPM Package would be fantastic, but any solution would work.

I've tried my best to get it to work but haven't gotten anywhere so far.

**Comment by dohomi** [comment]
I am making use of the new rewrites feature of NextJS v9.1.4
```js
//next.config.js
 experimental: {
      modern: true,
      async rewrites () {
        return [
          {source: '/sitemap.xml', destination: '/api/sitemap'},
        ]
      },
      catchAllRouting: true
    },
```
Inside of `pages/api/sitemap.ts` I run the npm module sitemap and build a dynamic sitemap on the fly based on content from a headless API:
```js
import { SitemapStream, streamToPromise } from 'sitemap'
import { IncomingMessage, ServerResponse } from 'http'

export default async function sitemapFunc(req: IncomingMessage, res: ServerResponse) {
  res.setHeader('Content-Type', 'text/xml')
  try {
    const stories =  await fetchContentFromAPI() // call the backend and fetch all stories
    
    const smStream = new SitemapStream({ hostname: 'https://' + req.headers.host })
    for (const story of stories) {
      smStream.write({
        url: story.full_slug,
        lastmod: story.published_at
      })
    }
    smStream.end()
    const sitemap = await streamToPromise(smStream)
      .then(sm => sm.toString())
    res.write(sitemap)
    res.end()
  } catch (e) {
    console.log(e)
    res.statusCode = 500
    res.end()
  }
}
```

**Comment by sj-log** [comment]
> I am making use of the new rewrites feature of NextJS v9.1.4
> 
> ```js
> //next.config.js
>  experimental: {
>       modern: true,
>       async rewrites () {
>         return [
>           {source: '/sitemap.xml', destination: '/api/sitemap'},
>         ]
>       },
>       catchAllRouting: true
>     },
> ```
> 
> Inside of `pages/api/sitemap.ts` I run the npm module sitemap and build a dynamic sitemap on the fly based on content from a headless API:
> 
> ```js
> import { SitemapStream, streamToPromise } from 'sitemap'
> import { IncomingMessage, ServerResponse } from 'http'
> 
> export default async function sitemapFunc(req: IncomingMessage, res: ServerResponse) {
>   res.setHeader('Content-Type', 'text/xml')
>   try {
>     const stories =  await fetchContentFromAPI() // call the backend and fetch all stories
>     
>     const smStream = new SitemapStream({ hostname: 'https://' + req.headers.host })
>     for (const story of stories) {
>       smStream.write({
>         url: story.full_slug,
>         lastmod: story.published_at
>       })
>     }
>     smStream.end()
>     const sitemap = await streamToPromise(smStream)
>       .then(sm => sm.toString())
>     res.write(sitemap)
>     res.end()
>   } catch (e) {
>     console.log(e)
>     res.statusCode = 500
>     res.end()
>   }
> }
> ```

I wonder why my `fetchContentFromAPI` gets error `Cannot find name 'fetchContentFromAPI'.`  . Do you know What should I do to find that function? Thank you in advance!

**Comment by dohomi** [comment]
@sj-log `fetchContentFromAPI` is a dummy function. You would need to write your own function to call your backend/service to fetch all links you would like to expose as a sitemap

**Comment by sj-log** [comment]
@dohomi Which function did you use for your case? Should I use 'getInitialProps()' at there?

**Comment by dohomi** [comment]
@sj-log I think you misunderstand: you can do whatever suits your usecase in this function. It is nothing specific to NextJS, it simply returns an array of items and I just gave it the name `fetchContentFromAPI`

**Comment by sj-log** [comment]
@dohomi Well Thank you for the answer, but I really have no idea how to fetch backend or the service as you mentioned.

**Comment by dohomi** [comment]
@sj-log what backend do you need to call? I am using a headless CMS (Storyblok). You either can use the native `fetch` or `Axios`, whatever suits your need

**Comment by MihaiWill** [comment]
@dohomi Perfect!!! Thanks a lot! What about:
``` 
Warning: You have enabled experimental feature(s).
Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.
```

**Comment by dohomi** [comment]
@MihaiWill you will see this warning as soon you enable any of the `experimental` features of NextJS. As it states: the experimental features are not considered stable and used by your own risk. I am running them very stable so far :-)

**Comment by sj-log** [comment]
@dohomi Dear friend, Just I've found another sitemap-generator for nextjs.
so I solved this site mapping issue! So sorry if I made you a bit frustrating! 

* if there is someone who is struggling to build a sitemap in nextjs, please refer to this one as well. [I got this.](https://github.com/IlusionDev/nextjs-sitemap-generator) 

dohomi's way to implement is quite simple and good, However, If you still feel 'I don't know what to do' status, please check out the above link. Good days.

**Comment by leerob** [comment]
I've spent some time looking into this. I didn't want to mess with a custom server, or utilize rewrites. Here's where I landed.

```js
// pages/sitemap.xml.js
import fetch from 'isomorphic-unfetch';
import React from 'react';

const EXTERNAL_DATA_URL = 'https://jsonplaceholder.typicode.com/posts';

const createSitemap = (posts) => `<?xml version="1.0" encoding="UTF-8"?>
    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
        ${posts
            .map(({id}) => {
                return `
                    <url>
                        <loc>${`${EXTERNAL_DATA_URL}/${id}`}</loc>
                    </url>
                `;
            })
            .join('')}
    </urlset>
    `;

class Sitemap extends React.Component {
    static async getInitialProps({res}) {
        const request = await fetch(EXTERNAL_DATA_URL);
        const posts = await request.json();

        res.setHeader('Content-Type', 'text/xml');
        res.write(createSitemap(posts));
        res.end();
    }
}

export default Sitemap;
```

[More info here.](https://masteringnextjs.com/sitemap-robots)

**Comment by perryraskin** [comment]
@leerob took a look at your blog post, really helpful! do you know what I would do to make a dynamic sitemap if my app creates blog post pages on the fly via [slug].tsx? For example, right now https://raskin.me/blog/init(blog) goes to my blog post, but it uses the template [slug].tsx to generate it. Currently the sitemap grabbed the actual "[slug]" name.

**Comment by leerob** [comment]
@perryraskin I think you would need to slightly modify your script to look at your `/posts` folder via something like `posts/*.md`.

```js
const pages = await globby(['posts/*.md', 'pages/**/*.tsx', '!pages/_*{.jsx,.tsx}']);
 ```

That will fetch all your Markdown files. Then, you'd need to probably replace `posts` with `blog` in the URL you output to your sitemap.

Give that a shot!

**Comment by perryraskin** [comment]
@leerob nice, that worked great. the sitemap is still showing the [slug] though, should i remove it? i tried doing sitemap.replace but it does not seem to be having any effect

**Comment by leerob** [comment]
@perryraskin You'll probably want to exclude the blog folder since you're handling it via `posts`. This should get it working for ya! 

```js
const pages = await globby(['posts/*.md', '!pages/blog', 'pages/**/*.tsx', '!pages/_*{.jsx,.tsx}']);
```

Another idea to consider: utilize `getStaticPaths` and `getStaticProps` that will be released [soon](https://github.com/zeit/next.js/issues/9524) to generate a static-site (SSG) instead of server-side rendering (SSR) your blog posts. Here's an [example](https://github.com/Timer/blog/blob/master/pages/post/%5Bslug%5D.tsx#L20).

Once SSG support lands, I'll probably update the blog post to include an example for that.

https://github.com/zeit/next.js/discussions/10437#discussioncomment-333

**Comment by pspeter3** [comment]
`getStaticPaths` and `getStaticProps` just launched. Can you use them to generate non HTML content (eg RSS / Sitemap)?

**Comment by leerob** [comment]
@pspeter3 After reading the [new docs](https://nextjs.org/docs/basic-features/data-fetching#statically-generates-both-html-and-json), it seems like `getStaticProps` generates HTML/JSON. To create non-HTML content, I believe you'd need to use [`getServerSideProps`](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering).

```js
export async function getServerSideProps({res}) {
    const request = await fetch(EXTERNAL_DATA_URL);
    const posts = await request.json();

    res.setHeader('Content-Type', 'text/xml');
    res.write(createSitemap(posts));
    res.end();
}
```

**Comment by pspeter3** [comment]
@leerob Thanks! I'm able to get that to run but I end up with `admin/config.yml.html` _(I'm_ trying to generate configuration for Netlify CMS)_. Do you how I can **not** generate the `.html` suffix?

**Comment by leerob** [comment]
@pspeter3 The way I've gotten around that myself is using `replace` for file extensions. I used this when mapping directly over the `pages` directory looking at MDX files. I believe a similar approach could work for you to filter out `.html` files after talking to Netlify CMS.

```js
const sitemap = `
    <?xml version="1.0" encoding="UTF-8"?>
    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
        ${pages
            .map((page) => {
                const path = page
                    .replace('pages', '')
                    .replace('.js', '')
                    .replace('.mdx', '');
                const route = path === '/index' ? '' : path;

                return `
                    <url>
                        <loc>${`https://yoursitehere.com${route}`}</loc>
                    </url>
                `;
            })
            .join('')}
    </urlset>
`;
```

**Comment by pspeter3** [comment]
Got it. I was actually commenting on something different. I was trying to use Next to generate a static file that does not have a `.html` ending. I think that really I want the ability to generate static files using the Next infrastructure and be able to set the permalink. Maybe I should open a new issue?

**Comment by leerob** [comment]
@pspeter3 Ahh, sorry for the confusion there – my bad. I agree, I would open up a new issue and try and provide a solid example 👍

**Comment by pspeter3** [comment]
Added #11115 as the new issue

**Comment by BrunoBernardino** [comment]
Here I have a solution that allows you to have a dynamically-generated sitemap that covers dynamic routes and static ones, as an interim solution of sorts (I've already commented on the issue above, but for those who land here and aren't following that one):

It uses a [lambda](https://github.com/BrunoBernardino/nextjs-boilerplate-faunadb-elasticsearch/blob/dd723c8d1066401a6a28989fffade0b25b0b509e/lambdas/sitemap.ts) for the sitemap generation, routed via [`@now/node` in `vercel.json`](https://github.com/BrunoBernardino/nextjs-boilerplate-faunadb-elasticsearch/blob/dd723c8d1066401a6a28989fffade0b25b0b509e/vercel.json#L33-L43), with the static pages/routes [built with a script](https://github.com/BrunoBernardino/nextjs-boilerplate-faunadb-elasticsearch/blob/dd723c8d1066401a6a28989fffade0b25b0b509e/scripts/build-pages.js) before deploy.

Maybe it'll help you. It supports static routes, and allows for newly-dynamic pages to still be valid and added (like for products, in that example).

**Comment by t-lochhead** [comment]
> I've spent some time looking into this. I didn't want to mess with a custom server, or utilize rewrites. Here's where I landed.
> 
> ```js
> // pages/sitemap.xml.js
> import React from 'react';
> 
> const EXTERNAL_DATA_URL = 'https://jsonplaceholder.typicode.com/posts';
> 
> const createSitemap = (posts) => `<?xml version="1.0" encoding="UTF-8"?>
>     <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
>         ${posts
>           .map(({ id }) => {
>             return `
>                     <url>
>                         <loc>${`${EXTERNAL_DATA_URL}/${id}`}</loc>
>                     </url>
>                 `;
>           })
>           .join('')}
>     </urlset>
>     `;
> 
> class Sitemap extends React.Component {
>   static async getServerSideProps({ res }) {
>     const request = await fetch(EXTERNAL_DATA_URL);
>     const posts = await request.json();
> 
>     res.setHeader('Content-Type', 'text/xml');
>     res.write(createSitemap(posts));
>     res.end();
>   }
> }
> 
> export default Sitemap;
> ```
> 
> [More info here.](https://leerob.io/blog/nextjs-sitemap-robots)

Thank you @leerob for putting this together. The exact code wasn't working for me, but got it to work through some small edits. Sharing in case anyone else ran into the same problem.

`// pages/sitemap.xml.js

const EXTERNAL_DATA_URL = "https://jsonplaceholder.typicode.com/posts";

const createSitemap = (posts) => `<?xml version="1.0" encoding="UTF-8"?>
    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
        ${posts
          .map(({ id }) => {
            return `
                    <url>
                        <loc>${`${EXTERNAL_DATA_URL}/${id}`}</loc>
                    </url>
                `;
          })
          .join("")}
    </urlset>
    `;

// remove component
export async function getServerSideProps({ res }) {
  const request = await fetch(EXTERNAL_DATA_URL);
  const posts = await request.json();

  res.setHeader("Content-Type", "text/xml");
  res.write(createSitemap(posts));
  res.end();
}

// add component here
export default () => null;
`

**Comment by pspeter3** [comment]
This is a cool hack! Will it work for exporting too?

**Comment by t-lochhead** [comment]
@pspeter3 Not sure! First time I tried this.

**Comment by xaun** [comment]
@leerob what's the go with generating a sitemap like that for thousands, or even millions of posts? Would that be recommended or beneficial?

**Comment by leerob** [comment]
@xaun If there are millions of posts, there are likely _multiple_ sitemaps stitched together.

**Comment by BrunoBernardino** [comment]
@xaun have you tried https://github.com/vercel/next.js/issues/9051#issuecomment-661110802 ? It should work for your case as well.

**Comment by dbertella** [comment]
I have tried something very similar to what you are suggesting and succeded to get a sitemap working in my local machine for both dynamic and static routes but in prod it looks like static routes aren't generated in `getServerSideProps`. I'm using globby package to get the list of the pages but I guess it's not possible to use it in production?

This is my attempt if it can help someone else, https://github.com/dbertella/cascinarampina/commit/efdfea09a91888a59de67d340b0bac025c4d5a7c#diff-2e0db62b38c8e7637d13c40a1a3034cb516d8899478112624a04b6f1f5c0abde
I for tonight added the pages list as a static list for now.

I was thinking will this work in a function at build time and write a static xml file instead of having a ssr page?

**Comment by BrunoBernardino** [comment]
@dbertella that's what I suggest in https://github.com/vercel/next.js/issues/9051#issuecomment-661110802 if you want to try that out.

**Comment by dbertella** [comment]
Hey @BrunoBernardino yours is a good idea. I'm not super familiar with the whole node environment but if I understand correctly what you are doing the only difference with my script is that you generate the list of pages at build time and that's cool indeed. Regarding the lamda it seems you are using an older version of next, would it work the same if you put the function in the api folder? Also @now/node seems to be deprecated in case will it work the same with @vercel/node? The real question is if you move the lamda in the api folder would you still need to install the additional package?

I will try to generate the list of static pages at build time btw, see if it work fine with my current implementation and let you know in any case.

**Comment by BrunoBernardino** [comment]
Awesome to hear it worked! I'm sure newer versions might need some tweaks, but should still work fine.

**Comment by dbertella** [comment]
I'm actually thinking at a different approach now and going static by default. The issue I have is that I'm trying to use es6 imports in a node.js script and it doesn't like it very much, but I'd probably go back to it when I have time to try to figure it out this is my next attempt https://github.com/dbertella/cascinarampina/pull/9/files

**Comment by loganstellway** [comment]
Hi all 👋🏼  

I am currently exploring the possibility of using the [`getStaticProps`](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation) method to build a static `page/sitemap.xml.tsx` page and leverage the `revalidate` property for [incremental static regeneration](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration). 

My site has a good amount of dynamic content, and it would be a heavy / expensive job to dynamically generate a `sitemap.xml` on every request. 

From what I am finding, this will probably require a [custom server](https://nextjs.org/docs/advanced-features/custom-server) and / or a [custom document](https://nextjs.org/docs/advanced-features/custom-document). 

<details>
<summary>My first attempt to write a custom `pages/_document.tsx`</summary>

I exported a `nonHtmlDocument` property in my `pages/sitemap.xml.tsx` page to be referenced here. 

`pages/sitemap.xml.tsx`
```typescript
import { GetStaticProps } from "next";
...

/**
 * Get static page properties
 *
 * @param GetStaticPropsContext props
 * @return Object
 */
export const getStaticProps: GetStaticProps<SitemapProps> = async () => {
    return {
        revalidate: 60 * 60 * 24, // 1 day
        props: {
            modified: "",
            nonHtmlDocument: true,
        },
    };
};

...
```

```typescript
import Document from "next/document";

class MyDocument extends Document<{ nonHtmlDocument?: boolean }> {
    static renderDocument(document, props) {
        return props?.__NEXT_DATA__?.props?.pageProps?.nonHtmlDocument ? (
            <>{props?.html}</>
        ) : (
            super.renderDocument(document, props)
        );
    }

    render() {
        return this.props?.__NEXT_DATA__?.props?.pageProps?.nonHtmlDocument ? (
            <>{this?.props?.html}</>
        ) : (
            <Document {...this.props} />
        );
    }
}

export default MyDocument;
```
</details>

Issues with my attempt:
- I still see the `<!DOCTYPE html>` rendered
- I need to output the `Content-Type` header

**Comment by lkbr** [comment]
Is there a particular reason why you've chosen getStaticProps over an /api route like was suggested here: https://github.com/vercel/next.js/issues/9051#issuecomment-556824393?

We've been happy with an /api/sitemap.ts route + a /sitemap.xml => /api/sitemap.xml rewrite + https://vercel.com/docs/edge-network/caching#stale-while-revalidate. 

It feels less of a hack than trying to turn a react page into a sitemap.

**Comment by loganstellway** [comment]
@lkbr 
Thank you for your response - 
I agree, the API route is a much cleaner solution. 

> Is there a particular reason why you've chosen getStaticProps over an /api route like was suggested here: [#9051 (comment)](https://github.com/vercel/next.js/issues/9051#issuecomment-556824393)?

My main concern was with incremental static regeneration, but I was oblivious to how that was implemented with the `Cache-Control` header.
_(I assumed Next.js implemented some sort of additional logic for this)_

Knowing this, I see that my CDN does support the `stale-while-revalidate` header and this approach should work great for me. 

Thank you 🙏🏼

**Comment by KishokanthJeganathan** [comment]
tips from everyone here really helped me. Here is how I did this with contenful CMS

```
import React from 'react';

const client = require('contentful').createClient({
  space: process.env.CONTENTFUL_SPACE,
  accessToken: process.env.CONTENTFUL_TOKEN,
});

const createSitemap = (
  allBlogPortfolioServiceEntries
) => `<?xml version="1.0" encoding="UTF-8"?>
    ${allBlogPortfolioServiceEntries
      .map((post) => {
        return `
      <url>
          <loc>https://kishokanth.com/${post.fields.category}/${post.fields.slug}</loc>
          <changefreq>daily</changefreq>
          <priority>0.7</priority>
      </url>
  `;
      })

      .join('')}
    </urlset>
    `;

export async function getServerSideProps({ res }) {
  const allEntries = await client.getEntries();

  const allBlogPortfolioServiceEntries = allEntries.items.filter(
    (item) =>
      item.sys.contentType.sys.id === 'servicesOffered' ||
      item.sys.contentType.sys.id === 'portfolioCaseStudies' ||
      item.sys.contentType.sys.id === 'blogPosts'
  );

  res.setHeader('Content-Type', 'text/xml');
  res.write(createSitemap(allBlogPortfolioServiceEntries));
  res.end();
  return {
    props: {}, // will be passed to the page component as props
  };
}

const sitemap = () => {
  return <div></div>;
};

export default sitemap;
```

**Comment by samos123** [comment]
I implemented @leerob solutions and it works however it's a bit hacky. Is there any plan for NextJS to have a better sitemap solution? There should be an official recommendation on sitemap generation in the docs because it's quite common. I've moved to NextJS mostly because of SEO. So supporting SEO use cases like sitemap generation are very important to me.

**Comment by loganstellway** [comment]
I was recently browsing through the [NextJS `examples` directory](https://github.com/vercel/next.js/tree/canary/examples) and came across the [`with-sitemap` example](https://github.com/vercel/next.js/tree/canary/examples/with-sitemap). 

The approach implements a [sitemap generation script](https://github.com/vercel/next.js/blob/canary/examples/with-sitemap/scripts/generate-sitemap.js) that outputs a static `sitemap.xml` file to the `public` directory. The script is included in the [`next.config.js` `webpack` extension](https://github.com/vercel/next.js/blob/cc541fb25260f7315d3e223d368cfad80c99f25a/examples/with-sitemap/next.config.js#L3-L5) ([Custom Webpack Config docs](https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config)). 

I plan to implement a similar sitemap generation script in my project and run it via a cron job (vs on build). This way, I will have more control over when the sitemap is generated & do not have to rely on any cache headers (like in my [previous approach](https://github.com/vercel/next.js/issues/9051#issuecomment-736073441)).

**Comment by BrunoBernardino** [comment]
@loganstellway that's an interesting idea. If you need dynamic content (that might not exist at build time), you could consider the solution I have in [this example repo](https://github.com/BrunoBernardino/nextjs-boilerplate-blog-markdown/blob/master/pages/sitemap.xml.ts).

**Comment by jupardo** [comment]
Tried using functions to generate sitemap, however my generated sitemap is huge so i cannot generate it with lambda. Is there other option of generating sitemap at build / call? Any help would be appreciated!

**Comment by BrunoBernardino** [comment]
@jupardo have you tried the solution I've suggested above?

**Comment by amala-james-cko** [comment]
@leerob, I tried [this](https://leerob.io/blog/nextjs-sitemap-robots) for both dynamic and static pages. but it looks like I have two sitemaps is there a way to combine both together?

**Comment by leerob** [comment]
We are adding [official docs on sitemaps](https://github.com/vercel/next.js/pull/28346), please leave any feedback there! 🙏

_P.S. I was going to clean this up and remove outdated solutions, but I'll leave it for posterity. Again, please share any and all feedback on the PR for adding this to the docs - Thank you!_
